subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c

```diff
@@ -527,22 +528,24 @@ static int get_tx_buffer(const struct device *instance, void *token,
 	}
 
 	/* The user requested a specific size */
-	if (*size) {
-		buf_size = get_tx_buffer_size(instance, token);
-		if (buf_size < 0) {
-			return -EIO;
-		}
-
+	if ((*size) && (*size > data->tx_buffer_size)) {
 		/* Too big to fit */
-		if (*size > buf_size) {
-			*size = buf_size;
-			return -ENOMEM;
-		}
+		*size = data->tx_buffer_size;
+		return -ENOMEM;
 	}
 
-	payload = rpmsg_get_tx_payload_buffer(&rpmsg_ept->ep, size, K_TIMEOUT_EQ(wait, K_FOREVER));
+	/*
+	 * OpenAMP doesn't really have the concept of forever but instead it
+	 * gives up after 15 seconds.  In that case, just keep retrying.
+	 */
+	do {
+		payload = rpmsg_get_tx_payload_buffer(&rpmsg_ept->ep, size,
+						      K_TIMEOUT_EQ(wait, K_FOREVER));
+	} while ((!payload) && K_TIMEOUT_EQ(wait, K_FOREVER));
```

Comment:
How does it work for a remote endpoint? According to the PR description I guess we would enter busy loop. Is my understanding correct?

Reply:
This loop is quite strange for me.
`K_TIMEOUT_EQ(wait, K_FOREVER) `return true if wait == K_FOREVER, else false, right?
In this case if user requests a time out of `x` milliseconds K_TIMEOUT_EQ will return false.

If the aim is to manage a busy loop with a timeout out, perhaps implementing a loop calling   `rpmsg_get_tx_payload_buffer` with wait set to false and  a `k_sleep` or a `k_yield` would be closer to the expected behavior.

Reply:
This looks good to me also, compared to previous approach. It supports now `_NO_WAIT` and `_FOREVER` timeout values. For the  `_FOREVER` the `rpmsg_get_tx_payload_buffer` can block for 1 seconds and check if there is free buffer up to 15 times. The Zephyr ports uses `k_sleep` function to wait for buffer so this is not exactly busy loop because context switching can be done.

I like also the @arnopo proposition which would add possibility to request any timeout value, it can be done under this PR but I also can agree that we can approach it the another PR

Reply:
> For the `_FOREVER` the `rpmsg_get_tx_payload_buffer` can block for 1 seconds and check if there is free buffer up to 15 times.

Does it work this way for both host and remote sides?

Reply:
@hubertmis  

> How does it work for a remote endpoint? According to the PR description I guess we would enter busy loop. Is my understanding correct?

Why the behavior on the remote endpoint should be different? AFAICT there is no difference between host and remote.

@arnopo 

> In this case if user requests a time out of x milliseconds K_TIMEOUT_EQ will return false.

The user cannot request anything different from `_NO_WAIT` and `_FOREVER`, see: https://github.com/zephyrproject-rtos/zephyr/blob/f2e2bbd8642b47a18434f410e0813e24162252d6/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L524-L527

> If the aim is to manage a busy loop with a timeout out, perhaps implementing a loop calling rpmsg_get_tx_payload_buffer with wait set to false and a k_sleep or a k_yield would be closer to the expected behavior.

I'm not trying to support a general timeout value in this PR.

@KAGA164 

> The Zephyr ports uses k_sleep function to wait for buffer so this is not exactly busy loop because context switching can be done here.

This is exactly it.

Reply:
> Why the behavior on the remote endpoint should be different? AFAICT there is no difference between host and remote.

I must have misunderstood the commit message. Thanks for clarification.

Reply:
> @arnopo
> 
> > In this case if user requests a time out of x milliseconds K_TIMEOUT_EQ will return false.
> 
> The user cannot request anything different from `_NO_WAIT` and `_FOREVER`, see:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/f2e2bbd8642b47a18434f410e0813e24162252d6/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L524-L527
> 
> > If the aim is to manage a busy loop with a timeout out, perhaps implementing a loop calling rpmsg_get_tx_payload_buffer with wait set to false and a k_sleep or a k_yield would be closer to the expected behavior.
> 
> I'm not trying to support a general timeout value in this PR.

clear now, thanks!